codaco/Network-Canvas
# Add 'live reload' option for running cordova

This adds a 'live reload' dev workflow for mobile devices & simulators. Pointing cordova at the running dev server ended up being more straightforward (and a better dev ux) than flushing a [modified] dev bundle to disk & syncing.

Testing:
- Run `npm start` as usual
- Run `npm dev:ios` or `npm dev:android` (see readme for more options)
- Change something in app source

To make this work, a few things are happening:
- Webpack dev server leaves a '.devserver' file while running, which contains the LAN URL that devices & simulators can use.
- Cordova hooks prepare a 'dev server' config.xml before run, and restore
the original once the app launches. (This is pretty quick on device or if an emulator happens to be running; slow otherwise.)

Once running, mobile apps pick up live changes just like other browsers.

This is only triggered for the `run` stage of cordova, and a `LIVE_RELOAD` env var must be set. I think that's sufficient to keep this away from dist builds.

Caveats/warnings:

The primary drawback here is that we actually have to modify config.xml on disk, as far as I could see from looking at cordova source. It's reset to the original once the `run` command completes. But, [warning] if you interrupt then there's nothing currently catching that case. And if you do something like `git commit -a` while your emulator is starting up, you could wind up with unwanted config.xml changes.

We can discuss approaches to make this more robust, but I didn't want to continue right now. It may be worth seeing how well this works for people first.

<strike>TODO: test Windows.</strike>

Fixes #350.
